src/shared/sam/cli/samCliDetection.ts

```diff
@@ -0,0 +1,116 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as AsyncLock from 'async-lock'
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import { extensionSettingsPrefix, samAboutInstallUrl } from '../../constants'
+import { DefaultSettingsConfiguration } from '../../settingsConfiguration'
+import { SamCliConfiguration } from './samCliConfiguration'
+import { DefaultSamCliLocationProvider } from './samCliLocator'
+
+const localize = nls.loadMessageBundle()
+const lock = new AsyncLock()
+
+const learnMore = localize(
+    'AWS.samcli.userChoice.visit.install.url',
+    'Get SAM CLI'
+)
+
+const browseToSamCli = localize(
+    'AWS.samcli.userChoice.browse',
+    'Locate SAM CLI...'
+)
+
+const settingsUpdated = localize(
+    'AWS.samcli.detect.settings.updated',
+    'Settings updated.'
+)
+
+const settingsNotUpdated = localize(
+    'AWS.samcli.detect.settings.not.updated',
+    'No settings changes necessary.'
+)
+
+export async function detectSamCli(showMessageIfDetected: boolean): Promise<void> {
+    await lock.acquire('detect SAM CLI', async () => {
```

Comment:
When `detectSamCli` is invoked while a detection is already in progress, the second invocation should NO-OP. With this implementation, it will be queued for after the first invocation completes. So we could still end up spamming the user.

Reply:
The intention is that after calling detectSamCli, a caller can immediately start using the sam cli location offered by SamCliConfiguration. For that reason, it should not no-op, however it should maybe nudge showMessageIfDetected to false.  Are we able to define "if waiting for lock" functionality?

Reply:
Sorry, 'NO-OP' was the wrong terminology. I meant that it shouldn't start a new promise, but return the existing one. NodeJS is single-threaded, so this is relatively simple:

```typescript
let detectSamCliPromise: Promise<void> | undefined
export async function detectSamCli(...): Promise<void> {
    if (!!detectSamCliPromise) {
        return detectSamCliPromise
    }

    detectSamCliPromise = async () => {
        // ...
        detectSamCliPromise = undefined
    }()
}
```